sdks/python/apache_beam/runners/worker/statesampler.pyx

```diff
@@ -180,21 +184,43 @@ cdef class StateSampler(object):
         self.scoped_states_by_index[self.current_state_index].name,
         self.state_transition_count)
 
-  def scoped_state(self, name):
-    """Returns a context manager managing transitions for a given state."""
-    cdef ScopedState scoped_state = self.scoped_states_by_name.get(name, None)
+  def scoped_state(self, step_name, state_name=None, io_target=None):
+    """Returns a context manager managing transitions for a given state.
+    Args:
+      step_name: A string with the name of the running step.
+      state_name: A string with the name of the state (e.g. 'process', 'start')
+      io_target: An IOTargetName object describing the io_target (e.g. writing
+        or reading to side inputs, shuffle or state). Will often be None.
+
+    Returns:
+      A ScopedState for the set of step-state-io_target.
+    """
+    cdef ScopedState scoped_state
+    if state_name is None:
+      # If state_name is None, the worker is still using old style
+      # msec counters.
+      counter_name = '%s-%s-msecs' % (self.prefix, step_name)
```

Comment:
<!--new_thread; commit:b3d629fb27b7150b79f1bfcea973f5ff0c0a1b78; resolved:0-->
Shouldn't this be `'%s%s-msecs'` to preserve the old behavior?

Also, there is some confusion in this code, in this legacy case where only `step_name` is provided, in what is used to index into `self.scoped_states_by_name` (i.e., is it `step_name` as you do on line 203, or is it `counter_name`, as done on line 223?).

It is also a little messy that `counter_name` can be either a string (this branch) or a `CounterName` object (the else branch)...

Reply:
1. You can see on [line 117](https://github.com/apache/beam/pull/3786/files#diff-47ba9d6e6336e5d52725d77c3bd2dbd2R117) that we're chopping off the dash from prefixes, so now we add the dash here.

2. You are completely right about the confusion. Good catch. I've updated the code to reflect this.

3. I understand. This is temporary while all the clients start providing state_name. This should take less than a couple weeks once this change is pulled in - and then the legacy path can be removed fully.